server/src/main/java/com/cloud/server/ManagementServerImpl.java

```diff
@@ -4484,8 +4486,20 @@ public String uploadCertificate(final UploadCustomCertificateCmd cmd) {
         final String certificate = cmd.getCertificate();
         final String key = cmd.getPrivateKey();
 
-        if (cmd.getPrivateKey() != null && !_ksMgr.validateCertificate(certificate, key, cmd.getDomainSuffix())) {
-            throw new InvalidParameterValueException("Failed to pass certificate validation check");
+        if (key != null) {
+            Pair<Boolean, String> result = _ksMgr.validateCertificate(certificate, key, cmd.getDomainSuffix());
+            if (!result.first()) {
+                throw new InvalidParameterValueException(String.format("Failed to pass certificate validation check with error: %s", result.second()));
+            }
+        } else {
+            try {
+                s_logger.debug(String.format("Trying to validate the root certificate format"));
+                CertificateHelper.buildCertificate(certificate);
+            } catch (CertificateException e) {
+                String errorMsg = String.format("Failed to pass certificate validation check with error: Certificate validation failed due to exception: %s", e.getMessage());
+                s_logger.error(errorMsg);
+                throw new InvalidParameterValueException(errorMsg);
+            }
         }
```

Comment:
can you factor this out into new method(s), please? (or disect the whole method)

Reply:
updated the PR with a new method @DaanHoogland